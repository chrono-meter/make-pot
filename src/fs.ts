import type { Args } from './types'
import path from 'path'
import fs from 'fs/promises'
import { generatePotHeader } from './utils'

export async function writePotFile (args: Args, fileContent: string): Promise<void> {
  // the path to the .pot file
  const potFilePath = path.join(args.destination, `${args.slug}.pot`)

  await fs.writeFile(potFilePath, generatePotHeader(args) + fileContent)
}
