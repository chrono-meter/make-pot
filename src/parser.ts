import type { Args, Patterns, TranslationStrings } from './types'
import cliProgress, { type SingleBar } from 'cli-progress'
import { parseFile } from './extractors'
import { allowedFiles } from './const'
import { Glob, Path } from 'glob'
import { minimatch } from 'minimatch'
import Parser from 'tree-sitter'

// @ts-expect-error
import Js from 'tree-sitter-javascript'
// @ts-expect-error
import Php from 'tree-sitter-php'
// @ts-expect-error
import Ts from 'tree-sitter-typescript'
import path from 'path'
import gettextParser, { GetTextTranslations } from 'gettext-parser'
import { generateHeaderComments } from './utils'
import { consolidate } from './consolidate'

/**
 * Return the parser based on the file extension
 *
 * @param file - Path to the file
 * @return {Parser|{}|null} - the parser to be used with the file or null if no parser is found
 */
export function getParser(file: string): string | Parser {
	const ext = file.split('.').pop()
	switch (ext) {
		case 'ts':
			return Ts.typescript
		case 'tsx':
			return Ts.tsx
		case 'js':
		case 'jsx':
		case 'mjs':
		case 'cjs':
			return Js
		case 'php':
			return Php
		default:
			return ext!
	}
}

/**
 * Determines if a pattern represents a file, a directory, or a glob pattern.
 * @param pattern - The pattern string to evaluate.
 * @returns 'file', 'directory', or 'glob'.
 */
export function detectPatternType(
	pattern: string
): 'file' | 'directory' | 'glob' {
	const containsFileExtension = pattern.includes('.')
	const containsDirectorySeparator =
		pattern.includes(path.sep) || pattern.endsWith(path.sep)

	if (pattern.includes('*')) {
		return 'glob'
	} else if (!containsFileExtension && !containsDirectorySeparator) {
		return 'directory'
	} else if (containsFileExtension && !containsDirectorySeparator) {
		return 'file'
	} else {
		return 'glob'
	}
}

export function includeFunction(includePath: string[]) {
	return includePath.map((path) => {
		const type = detectPatternType(path)
		switch (type) {
			case 'directory':
				return '**/' + path + '/**'
			case 'file':
				return '**/' + path
			default:
				return path
		}
	})
}

/**
 * Retrieves a list of files based on the provided arguments and patterns.
 *
 * @param {Args} args - The argument object containing the source directory and other options.
 * @param {Patterns} pattern - The pattern object containing the included and excluded file patterns.
 * @return A promise that resolves to an array of file paths.
 */
export async function getFiles(args: Args, pattern: Patterns) {
	const includedPatterns = includeFunction(
		pattern.include.map((p) => p.trim()).filter((p) => p)
	) ?? ['**']

	// Process excludePaths
	const excludedPatterns = pattern.exclude ?? []

	// Build the ignore function for Glob
	const ignoreFunc = (filePath: Path): boolean => {
		return excludedPatterns.some((exclude) => {
			const type = detectPatternType(exclude)
			switch (type) {
				case 'file':
					return filePath.name === exclude
				case 'directory':
					return filePath.path.includes(exclude)
				default:
					// Handle glob patterns using minimatch or a similar library
					return minimatch(filePath.path, exclude)
			}
		}) as boolean
	}

	if (!args.silent)
		console.log(
			'Searching in :',
			path.resolve(args.sourceDirectory),
			'for ' + includeFunction(includedPatterns).join(),
			'\nExcluding : ' + excludedPatterns.join()
		)

	// Execute the glob search with the built patterns
	return new Glob(includedPatterns, {
		ignore: {
			ignored: (p) => {
				return ignoreFunc(p)
			},
		},
		nodir: true,
		cwd: args.sourceDirectory,
	})
}

/**
 * Initializes a progress bar and returns the progress bar element.
 *
 * @param {Args} args - The argument object containing the source directory and other options.
 * @param {number} filesCount - An array of file names.
 * @return {cliProgress.SingleBar} The progress bar element.
 */
function initProgress(args: Args, filesCount: number): SingleBar | null {
	if (args.silent) return null
	// Set up the progress bar
	const progressBar = new cliProgress.SingleBar(
		{
			clearOnComplete: true,
			etaBuffer: 1000,
			hideCursor: true,
			format: ' {bar} {percentage}% | ETA: {eta}s | {filename} | {value}/{total}',
		},
		cliProgress.Presets.shades_classic
	)

	progressBar.start(filesCount, 0)

	// Return the progress bar element
	return progressBar
}

/**
 * Retrieves an array of translation strings from files that match the specified arguments and pattern.
 *
 * @param {Args} args - The arguments to specify which files to search and parse.
 * @param {string[]} files - An array of file names to search and parse.
 * @return {Promise<TranslationStrings[]>} A promise that resolves to an array of translation strings found in the files.
 */
export async function getStrings(
	args: Args,
	files: Glob<{}>
): Promise<TranslationStrings> {
	const tasks: Promise<TranslationStrings>[] = []

	const progressBar = initProgress(
		args,
		Array.from(files.iterateSync()).length
	)

	// loop through the files and parse them
	for (const file of files) {
		// get the file extension
		const ext = file.split('.').pop() || 'undefined'
		// check if the extension is allowed
		if (!allowedFiles.includes(ext)) {
			// log the filepath
			if (progressBar) {
				progressBar.increment(1, {
					filename: `Skipping ${ext} (not a valid file extension)`,
				})
			}
			continue
		}

		const task = parseFile({
			filepath: path.join(process.cwd(), args.sourceDirectory, file),
			language: getParser(file),
		}) as Promise<TranslationStrings>

		// log the filepath
		if (progressBar) {
			progressBar.increment(1, {
				filename: file,
			})
		}

		// add the task to the array if it's not null
		if (task !== null) tasks.push(task as Promise<TranslationStrings>)
	}

	const results = await Promise.all(tasks)

	// stop the progress bar if it's not silent
	if (progressBar) progressBar.stop()

	const mergedResult = consolidate(results.filter((r) => r !== null).flat())

	console.log('Strings grouped')

	if (!args.silent) {
		console.log(
			'📝 Found',
			Object.values(mergedResult).length,
			'group of strings in',
			results.length,
			'files.\n',
			'In total ' +
				Object.values(mergedResult)
					.map((v) => Object.keys(v).length)
					.reduce((acc, val) => acc + val, 0) +
				' strings were found'
		)
	}

	return mergedResult
}

/**
 * Runs the extract process based on the given arguments.
 *
 * @param {Args} args - The arguments for the extract process.
 * @return {Promise<string>} - A promise that resolves with the extracted data.
 */
export async function runExtract(args: Args) {
	const pattern = {
		include: args.include || [],
		exclude: args.exclude || [],
		mergePaths: args.mergePaths,
		subtractPaths: args.subtractPaths,
		subtractAndMerge: args.subtractAndMerge,
	} as Patterns

	// Additional logic to handle different file types and formats
	// Exclude blade.php files if --skip-blade is set
	if (args.skipPhp !== undefined || args.skipBlade !== undefined) {
		if (args.skipBlade !== undefined) {
			// php files but not blade.php
			pattern.include.push('**/*.php')
		} else {
			pattern.include.push('**/*.php', '!**/blade.php')
		}
	}

	// js typescript mjs cjs etc
	if (args.skipJs !== undefined) {
		pattern.include.push('**/*.{js,jsx,ts,tsx,mjs,cjs}')
	}

	if (args.skipBlockJson !== undefined) {
		pattern.include.push('block.json')
	}

	if (args.skipThemeJson !== undefined) {
		pattern.include.push('theme.json')
	}

	const files = await getFiles(args, pattern)
	const stringsJson = await getStrings(args, files)

	const additionalHeaders = {
		'content-type': 'text/plain; charset=iso-8859-1',
		...args.headers,
		'plural-forms': 'nplurals=2; plural=(n != 1);',
	}

	const buffer = gettextParser.po.compile(
		{
			headers: additionalHeaders,
			comments: generateHeaderComments(args),
			translations: stringsJson,
			charset: 'iso-8859-1',
		} as unknown as GetTextTranslations,
		{
			sort: true,
		}
	)

	return buffer.toString('utf-8') as string
}
