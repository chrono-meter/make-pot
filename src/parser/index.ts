import path from "node:path";
import { extractMainFileData } from "../extractors/headers.js";
import { extractPackageJson } from "../extractors/json.js";
import { writeFile } from "../fs/index.js";
import type { Args } from "../types.js";
import { exec } from "./exec.js";

function getOutputPath(args: Args): string {
	return path.join(
		process.cwd(),
		args.headers?.domainPath ?? args.paths.out ?? "languages",
		`${args?.slug}.${args.options?.json ? "json" : "pot"}`,
	);
}

/**
 * Generates a pot file for localization.
 *
 * @param args - the command line arguments
 * @return {Promise<void>} - a promise that resolves when the pot file is generated
 */
export async function makePot(args: Args): Promise<string> {
	/** Collect metadata from the get package json */
	const pkgData = extractPackageJson(args);

	/** Get metadata from the main file (theme and plugin) */
	const metadata = extractMainFileData(args);

	/** Merge the metadata to get a single object with all the headers */
	args.headers = {
		...args.headers,
		...pkgData,
		...metadata,
	} as Args["headers"];

	/** Generate the pot file */
	const jsonTranslations = await exec(args);

	const outputPath = getOutputPath(args);

	console.log(`Writing pot file to ${outputPath}`);

	writeFile(jsonTranslations, outputPath);

	return jsonTranslations;
}
