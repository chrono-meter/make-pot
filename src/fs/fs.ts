import fs, { writeFileSync } from "node:fs";
import { readFile } from "node:fs/promises";
import path from "node:path";
import type { Args } from "../types";

/**
 * Ensures that a folder exists at the specified path.
 *
 * @param {string | undefined} folderPath - The path of the folder to ensure existence for.
 * @return {string} - The path of the folder, or '.' if folderPath is undefined.
 */
function ensureFolderExists(folderPath: string | undefined): string {
	if (folderPath === undefined) {
		return ".";
	}
	try {
		// Check if the folder exists
		fs.accessSync(
			path.resolve(folderPath),
			fs.constants.R_OK | fs.constants.W_OK,
		);
	} catch (error) {
		if ((error as NodeJS.ErrnoException).code === "ENOENT") {
			// The Folder does not exist, so create it
			fs.mkdirSync(folderPath, { recursive: true });
			console.log(`Folder created: ${folderPath}`);
			return folderPath;
		}
	}
	return folderPath;
}

/**
 * Gets the charset of the .pot file
 *
 * @param charset the charset of the .pot file
 * @return the charset of the .pot file
 */
export function getCharset(charset: string | undefined): BufferEncoding {
	if (!charset) {
		return "latin1";
	}
	// we need to check if the charset is valid otherwise we return latin1 that is a common alias for ISO-8859-1 and the default charset for pot files
	switch (charset.toLowerCase()) {
		case "utf-8":
			return "utf-8";
		default:
			return "latin1";
	}
}

export function getEncodingCharset(charset: string | undefined): string {
	if (!charset) {
		return "iso-8859-1";
	}
	// we need to check if the charset is valid otherwise we return utf-8 that is a common alias for ISO-8859-1 and the default charset for pot files
	switch (charset.toLowerCase()) {
		case "utf-8":
			return "utf-8";
		default:
			return "iso-8859-1";
	}
}

/**
 * The output path for the pot file.
 * @param outpath - the output path for the pot/json files
 * @return {string} - the output path
 */
export function getOutputPath(outpath?: string): string {
	return path.join(process.cwd(), outpath ?? "languages");
}

/**
 * The output path for the pot file.
 * @param args - the command line arguments
 */
function getOutputFilePath(args: Args): string {
	const { out, headers, options } = args.paths;
	let i18nFolder = out ?? headers?.domainPath ?? "languages";

	// Remove leading and trailing slashes
	i18nFolder = i18nFolder.replace(/^\/+|\/+$/g, "");

	const extension = options?.json ? "json" : "pot";

	return path.join(process.cwd(), i18nFolder, `${args.slug}.${extension}`);
}

/**
 * Writes the .pot file to disk
 *
 * @param fileContent the content of the .pot file
 * @param args the command line arguments passed to the program
 */
export function writeFile(fileContent: string, args: Args): void {
	const dest = getOutputFilePath(args);

	if (ensureFolderExists(path.dirname(dest))) {
		// get the encoding charset
		const encodingCharset = getCharset(args.options?.charset);
		console.log(`File created at ${dest}`);

		// write the file
		const potBuffer = Buffer.from(fileContent);
		writeFileSync(dest, potBuffer.toString(encodingCharset), {
			encoding: encodingCharset,
		});
	} else {
		console.log(`Folder ${dest} does not exist and cannot be created`);
	}
}

/**
 * The async version of fs.readFile method
 * @param path the path of the file to read
 * @return the content of the file as a string
 */
export async function readFileAsync(path: string): Promise<string> {
	return readFile(path, "utf-8");
}
