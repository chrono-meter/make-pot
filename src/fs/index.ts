import * as path from 'path'
import fs from 'node:fs'

/**
 * Ensures that a folder exists at the specified path.
 *
 * @param {string | undefined} folderPath - The path of the folder to ensure existence for.
 * @return {string} - The path of the folder, or '.' if folderPath is undefined.
 */
function ensureFolderExists(folderPath: string | undefined): string {
	if (folderPath === undefined) {
		return '.'
	}
	try {
		// Check if the folder exists
		fs.accessSync(
			path.resolve(folderPath),
			fs.constants.R_OK | fs.constants.W_OK
		)
	} catch (error) {
		if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
			// The Folder does not exist, so create it
			fs.mkdirSync(folderPath, { recursive: true })
			console.log(`Folder created: ${folderPath}`)
			return folderPath
		}
	}
	return folderPath
}

/**
 * Writes the .pot file to disk
 *
 * @param fileContent the content of the .pot file
 * @param dest the path of the .pot file to write
 */
export function writeFile(fileContent: string, dest: string) {
	if (ensureFolderExists(path.dirname(dest))) {
		fs.writeFileSync(dest, fileContent)
	}
}

export async function readFileAsync(path: string): Promise<string> {
	return fs.promises.readFile(path, 'utf-8')
}
