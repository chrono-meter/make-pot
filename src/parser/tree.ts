import Parser, { type SyntaxNode } from 'tree-sitter'
import { type TranslationStrings } from '../types'
import { i18nFunctions } from '../const'

import { GetTextComment, GetTextTranslation } from 'gettext-parser'
import { getParser } from '../fs/glob'
import { stripTranslationMarkup } from '../utils'

/**
 * Collect comments from the AST node and its preceding siblings.
 *
 * @param {SyntaxNode} node - The AST node.
 * @return {string[]} An array of collected comments.
 */
function collectComments(node: SyntaxNode): string | undefined {
	let currentNode = node
	let depth = 0

	// Check the node's preceding siblings for comments
	while (currentNode && depth < 6) {
		if (
			currentNode?.previousSibling?.type === 'comment' &&
			currentNode?.previousSibling?.text
				.toLowerCase()
				.includes('translators')
		) {
			return stripTranslationMarkup(currentNode?.previousSibling.text)
		}
		depth++
		currentNode = currentNode.parent as SyntaxNode
	}
}

/**
 * Parses the source code using the specified language parser and extracts the strings from the file.
 *
 * @param {string} sourceCode - The source code to be parsed.
 * @param {string} filepath - The path to the file being parsed.
 * @return {TranslationStrings[]} An array of translation strings.
 */
export function doTree(
	sourceCode: string,
	filepath: string
): TranslationStrings {
	// set up the parser
	const parser = new Parser()
	parser.setLanguage(getParser(filepath))

	// parse the file
	const tree = parser.parse(sourceCode)

	const gettextTranslations: TranslationStrings = {}
	const typeToMatch =
		filepath.split('.').pop()?.toLowerCase() !== 'php'
			? 'call_expression'
			: 'function_call_expression'

	/**
	 * Traverse the tree ðŸŒ³
	 *
	 * @param {SyntaxNode} node The node to traverse through
	 */
	function traverse(node: SyntaxNode): void {
		// Walk the tree
		if (node?.children.length)
			for (const child of node.children) {
				traverse(child)
			}

		// Check if the node matches
		if (node?.type === typeToMatch) {
			// The function name is the first child
			const functionName = node.firstChild?.text ?? null
			if (
				functionName === null ||
				!Object.keys(i18nFunctions).includes(functionName)
			) {
				return
			}

			// The arguments are the last child
			const argsNode = node.lastChild
			if (
				argsNode === null ||
				argsNode.childCount === 0 ||
				argsNode.type !== 'arguments'
			) {
				return
			}

			// Get the whole gettext translation string
			// eslint-disable-next-line @typescript-eslint/no-unused-vars
			const [_fn, raw] = node.children
			const translation: Partial<GetTextTranslation> = {}

			const translationKeys =
				i18nFunctions[functionName as keyof typeof i18nFunctions]

			const children = raw.children.slice(1, -1)
			let translationKeyIndex = 0

			// Get the translation from the arguments (the quoted strings)
			for (const child of children) {
				let node = child
				let nodeValue = node.text

				// unwrap the argument node, which is used in PHP.
				if (child.type === 'argument') {
					if (child.children.length === 0) continue
					node = child.children[0]
				}

				if (node?.type === ',') {
					// skip the comma between arguments
					continue
				} else if (
					node?.type === 'string' ||
					node.type === 'encapsed_string'
				) {
					// unquote the strings
					nodeValue = nodeValue.slice(1, -1)
				}

				// the translation key (eg. msgid)
				const currentKey = translationKeys[translationKeyIndex]

				// the value of that key
				translation[currentKey as keyof typeof translation] = nodeValue

				// increment the index of the translation key
				translationKeyIndex += 1
			}

			// Get the translation data
			const gettext: GetTextTranslation = {
				msgctxt: translation.msgctxt ?? '',
				msgid: translation.msgid ?? '',
				msgid_plural: translation.msgid_plural ?? '',
				msgstr: translation.msgid_plural ? ['', ''] : [''],
				comments: {
					translator: collectComments(node) ?? '',
					reference: `${filepath}:${node.startPosition.row + 1}`,
				} as GetTextComment,
			}

			gettextTranslations[gettext.msgctxt ?? ''] = {
				...(gettextTranslations[gettext.msgctxt ?? ''] || {}),
				[gettext.msgid]: gettext,
			}
		}
	}

	traverse(tree.rootNode)

	// Return both matches and entries
	return gettextTranslations
}
