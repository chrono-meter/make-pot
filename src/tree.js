#!/usr/bin/env node
import Parser from 'tree-sitter'
import Ts from 'tree-sitter-typescript'
import Js from 'tree-sitter-javascript'
import Php from 'tree-sitter-php'
import path from 'path'
import { glob } from 'glob'
import { readFile } from 'fs/promises'

/**
 * Extract strings from a file ðŸŒ³
 *
 * @param {Tree} node
 * @param {string} lang
 * @param {string} filename
 * @return {{}}
 */
function extractStrings (node, lang, filename) {
  const matches = []
  const typeToMatch = lang !== Php ? 'call_expression' : 'function_call_expression'

  /**
   * Traverse the tree ðŸŒ³
   *
   * @param {Tree} node The node to traverse through
   * @param {Tree|null} previousNode The previous node in the tree
   */
  function traverse (node, previousNode = null) {
    if (node.type === typeToMatch) {
      const functionNameNode = node.firstChild
      let meta = {}
      const functionName = functionNameNode !== null ? functionNameNode.text : null

      if (functionName?.match(/^(_n|_x|_nx|__|_e)$/) !== null) {
        const argsNode = node.lastChild
        const [fnName, ...msgid] = node.children.map(argNode => argNode.text)

        // Extract the metadata
        if (argsNode !== null && argsNode.childCount > 0 && argsNode.type === 'arguments') {
          // Extract the msgid
          const firstArgNode = argsNode.firstChild
          const msgidNode = firstArgNode.firstChild
          const msgid = msgidNode?.text

          // Extract the msgctxt if it exists and unquote it
          const msgctxtNode = firstArgNode.nextNamedSibling
          let msgctxt
          if (msgctxtNode !== null && msgctxtNode.type === 'string') {
            // unquote the string
            msgctxt = msgctxtNode.text.slice(1, -1)
          }

          let comments
          // Find the preceding comment node at the same level as the translation function call.
          let nodeToCheck = previousNode
          while (nodeToCheck !== null && nodeToCheck.type !== 'comment') {
            nodeToCheck = nodeToCheck.previousSibling
          }
          // If the node is a comment, extract the comment text
          if (nodeToCheck !== null && nodeToCheck.type === 'comment') {
            const commentText = nodeToCheck.text.trim()
            if (commentText.includes('translators:') !== false) {
              comments = commentText
            }
          }

          const position = node.startPosition
          meta = {
            msgid,
            msgctxt,
            comments,
            reference: `#: ${filename}:${position.row + 1}:${position.column + 1}` // Adjusted to be 1-indexed
          }
        }

        matches.push({ fnName, ...meta, msgid })
      }
    }

    if (node.children.length > 0) {
      for (const child of node.children) {
        // Make sure to pass the current node as previousNode to the next level of recursion.
        traverse(child, node)
      }
    }
  }

  traverse(node)

  // Return both matches and entries
  return matches
}

/**
 * Parse a file and extract strings
 *
 * @param {object} args
 * @param {string} args.filepath - Path to the file to parse
 * @param {string} args.language - Language of the file to parse
 * @return {{}}
 */
async function parseFile (args) {
  const sourceCode = await readFile(path.resolve(args.filepath), 'utf8')

  if (args.language === 'Text') {
    return { text: sourceCode }
  } else if (args.language === 'Json') {
    return JSON.parse(sourceCode)
  }

  const parser = new Parser()
  parser.setLanguage(args.language)
  const tree = parser.parse(sourceCode)

  return extractStrings(tree.rootNode, args.language, args.filepath) // Assuming extractStrings is synchronous
}

/**
 * Return the parser based on the file extension
 * @param file - Path to the file
 * @return {*|{}|string} - Parser
 */
function getParser (file) {
  const ext = file.split('.').pop()
  switch (ext) {
    case 'ts':
      return Ts.typescript
    case 'tsx':
      return Ts.tsx
    case 'js':
    case 'jsx':
    case 'mjs':
    case 'cjs':
      return Js
    case 'php':
      return Php
    case 'json':
      return 'Json'
    default:
      return 'Text'
  }
}

async function getFiles (src = 'sourcedir/**') {
  return await glob(src, { nodir: true, cwd: process.cwd() })
}

async function run () {
  const files = await glob('sourcedir/**', { nodir: true, cwd: process.cwd() })
  const parsePromises = files.map(async (file) => {
    console.log('Searching for strings in ' + file)
    const parsed = await parseFile({ filepath: file, language: getParser(file) })
    console.log(parsed)
    return parsed
  })

  const results = await Promise.all(parsePromises)
  console.log(results.flat())
}
const startTime = Date.now()
await run()
console.log('ðŸŽ‰ Done in', Date.now() - startTime, 'ms.')
