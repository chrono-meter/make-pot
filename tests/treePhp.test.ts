import { describe, expect } from '@jest/globals'
import { doTree } from '../src/parser/tree'
import path from 'path'
import fs from 'fs'

describe('doTree php', () => {
	it('Should build pot file', () => {
		const filePath = path.join(
			process.cwd(),
			'tests/fixtures/sourcedir/file.php'
		)
		const fileContent = fs.readFileSync(filePath, 'utf8')
		console.log('My file path is: ' + filePath)
		const fileParsed = doTree(fileContent, 'tests/fixtures/php.php')

		expect(fileParsed).toMatchSnapshot()
	})

	it('Should build pot file php', () => {
		const fileContent = fs.readFileSync(
			path.join(process.cwd(), 'tests/fixtures/sourcedir/file.php'),
			'utf8'
		)
		const fileParsed = doTree(
			fileContent,
			'tests/fixtures/sourcedir/php.php'
		)

		expect(fileParsed).toMatchSnapshot()
	})
})
