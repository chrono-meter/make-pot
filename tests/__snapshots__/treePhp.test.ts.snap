// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doTree php Should build pot file 1`] = `
{
  "": {
    "Huge": {
      "comments": {
        "reference": "tests/fixtures/php.php:252",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "Huge",
      "msgid_plural": "",
      "msgstr": [],
    },
    "Large": {
      "comments": {
        "reference": "tests/fixtures/php.php:247",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "Large",
      "msgid_plural": "",
      "msgstr": [],
    },
    "Light green cyan to vivid green cyan": {
      "comments": {
        "reference": "tests/fixtures/php.php:189",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "Light green cyan to vivid green cyan",
      "msgid_plural": "",
      "msgstr": [],
    },
    "Luminous vivid amber to luminous vivid orange": {
      "comments": {
        "reference": "tests/fixtures/php.php:194",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "Luminous vivid amber to luminous vivid orange",
      "msgid_plural": "",
      "msgstr": [],
    },
    "Luminous vivid orange to vivid red": {
      "comments": {
        "reference": "tests/fixtures/php.php:199",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "Luminous vivid orange to vivid red",
      "msgid_plural": "",
      "msgstr": [],
    },
    "Regular": {
      "comments": {
        "reference": "tests/fixtures/php.php:242",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "Regular",
      "msgid_plural": "",
      "msgstr": [],
    },
    "Small": {
      "comments": {
        "reference": "tests/fixtures/php.php:237",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "Small",
      "msgid_plural": "",
      "msgstr": [],
    },
    "Vivid cyan blue to vivid purple": {
      "comments": {
        "reference": "tests/fixtures/php.php:179",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "Vivid cyan blue to vivid purple",
      "msgid_plural": "",
      "msgstr": [],
    },
    "Vivid green cyan to vivid cyan blue": {
      "comments": {
        "reference": "tests/fixtures/php.php:184",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "Vivid green cyan to vivid cyan blue",
      "msgid_plural": "",
      "msgstr": [],
    },
    "light grayish magenta": {
      "comments": {
        "reference": "tests/fixtures/php.php:123",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "light grayish magenta",
      "msgid_plural": "",
      "msgstr": [],
    },
    "sdasdasdasd": {
      "comments": {
        "reference": "tests/fixtures/php.php:7",
        "translator": "ciao2!",
      },
      "msgctxt": "",
      "msgid": "sdasdasdasd",
      "msgid_plural": "",
      "msgstr": [],
    },
    "strong magenta": {
      "comments": {
        "reference": "tests/fixtures/php.php:118",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "strong magenta",
      "msgid_plural": "",
      "msgstr": [],
    },
    "very dark gray": {
      "comments": {
        "reference": "tests/fixtures/php.php:133",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "very dark gray",
      "msgid_plural": "",
      "msgstr": [],
    },
    "very light gray": {
      "comments": {
        "reference": "tests/fixtures/php.php:128",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "very light gray",
      "msgid_plural": "",
      "msgstr": [],
    },
  },
}
`;

exports[`doTree php Should build pot file php 1`] = `
{
  "": {
    "Huge": {
      "comments": {
        "reference": "tests/fixtures/sourcedir/file.php:252",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "Huge",
      "msgid_plural": "",
      "msgstr": [],
    },
    "Large": {
      "comments": {
        "reference": "tests/fixtures/sourcedir/file.php:247",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "Large",
      "msgid_plural": "",
      "msgstr": [],
    },
    "Light green cyan to vivid green cyan": {
      "comments": {
        "reference": "tests/fixtures/sourcedir/file.php:189",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "Light green cyan to vivid green cyan",
      "msgid_plural": "",
      "msgstr": [],
    },
    "Luminous vivid amber to luminous vivid orange": {
      "comments": {
        "reference": "tests/fixtures/sourcedir/file.php:194",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "Luminous vivid amber to luminous vivid orange",
      "msgid_plural": "",
      "msgstr": [],
    },
    "Luminous vivid orange to vivid red": {
      "comments": {
        "reference": "tests/fixtures/sourcedir/file.php:199",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "Luminous vivid orange to vivid red",
      "msgid_plural": "",
      "msgstr": [],
    },
    "Regular": {
      "comments": {
        "reference": "tests/fixtures/sourcedir/file.php:242",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "Regular",
      "msgid_plural": "",
      "msgstr": [],
    },
    "Small": {
      "comments": {
        "reference": "tests/fixtures/sourcedir/file.php:237",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "Small",
      "msgid_plural": "",
      "msgstr": [],
    },
    "Vivid cyan blue to vivid purple": {
      "comments": {
        "reference": "tests/fixtures/sourcedir/file.php:179",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "Vivid cyan blue to vivid purple",
      "msgid_plural": "",
      "msgstr": [],
    },
    "Vivid green cyan to vivid cyan blue": {
      "comments": {
        "reference": "tests/fixtures/sourcedir/file.php:184",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "Vivid green cyan to vivid cyan blue",
      "msgid_plural": "",
      "msgstr": [],
    },
    "light grayish magenta": {
      "comments": {
        "reference": "tests/fixtures/sourcedir/file.php:123",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "light grayish magenta",
      "msgid_plural": "",
      "msgstr": [],
    },
    "sdasdasdasd": {
      "comments": {
        "reference": "tests/fixtures/sourcedir/file.php:7",
        "translator": "ciao2!",
      },
      "msgctxt": "",
      "msgid": "sdasdasdasd",
      "msgid_plural": "",
      "msgstr": [],
    },
    "strong magenta": {
      "comments": {
        "reference": "tests/fixtures/sourcedir/file.php:118",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "strong magenta",
      "msgid_plural": "",
      "msgstr": [],
    },
    "very dark gray": {
      "comments": {
        "reference": "tests/fixtures/sourcedir/file.php:133",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "very dark gray",
      "msgid_plural": "",
      "msgstr": [],
    },
    "very light gray": {
      "comments": {
        "reference": "tests/fixtures/sourcedir/file.php:128",
        "translator": undefined,
      },
      "msgctxt": "",
      "msgid": "very light gray",
      "msgid_plural": "",
      "msgstr": [],
    },
  },
}
`;
